psh/test-date.py

```diff
@@ -0,0 +1,197 @@
+#
+# Copyright 2022, 2023 Phoenix Systems
+# Author: Mateusz Niewiadomski, Mateusz Bloch
+#
+# This file is part of Phoenix-RTOS.
+#
+# %LICENSE%
+#
+
+import psh.tools.psh as psh
+
+DFLT_DATE_REG = r"\w{3}, \d{2} \w{3} (?:\d{2}|\d{4}),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n"
+SPEC_DATE_REG = (
+    r"Sun, 13 Sep (?:20|2020),? 12:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n"  # arbitrary deta of epoch 1600000000 GMT
+)
+
+
+# Utility functions
+def invalid_format(format):
+    return f"date: invalid format '{str(format)}'"
+
+
+def invalid_date(format):
+    return f"date: invalid date '{format}'"
+
+
+def unrec_arg(arg):
+    return f"Unrecognized argument: {arg}"
+
+
+# Test parts
+def test_corrects(p):
+    """Testing correct `date` commands"""
+
+    # Test help
+    psh.assert_cmd(p, "date -h", expected=r"Usage(.*?)FORMAT(.*?)\r+\n", is_regex=True)
+
+    # Test printing and formatting
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+    psh.assert_cmd(p, "date +%Y", expected=r"\b(19|20)\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date +%H:%M:%Sformat", expected=r"\d{2}:\d{2}:\d{2}format\r+\n", is_regex=True)
+
+    # Test setting to value
+    psh.assert_cmd(p, "date -s @1600000000", expected=SPEC_DATE_REG, is_regex=True)
+    psh.assert_cmd(p, "date", expected=SPEC_DATE_REG, is_regex=True)
+
+    # Test setting to 0
+    psh.assert_cmd(p, "date -s @0", expected=r"Thu, 01 Jan (?:70|1970),? 00:00:00(?:\w{3}|\w{4})?\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=r"Thu, 01 Jan (?:70|1970),? 00:00:\d{2}(?:\w{3}|\w{4})?\r+\n", is_regex=True)
+
+    # Test date parsing
+    psh.assert_cmd(p, "date -d @1600000000", expected=SPEC_DATE_REG, is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+    psh.assert_cmd(p, "date -d @0", expected=r"Thu, 01 Jan (?:70|1970),? 00:00:00(?:\w{3}|\w{4})?\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_incorrect_dateprint(p):
+    """Test incorrect or rare commandlines for printing date"""
+
+    # Incorrect FORMAT passed when printing date
+    psh.assert_cmd(p, "date operand1", expected=invalid_format("operand1"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # Incorrect FORMAT passed when printing date
+    psh.assert_cmd(p, "date +operand1", expected="operand1")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # too many arguments passed to print, should print first redundant
+    psh.assert_cmd(p, "date operand1 operand2 operand3", expected=unrec_arg("operand2"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # nonexistent format '%k' passed to print
+    psh.assert_cmd(p, "date +%Y%k%Y", expected=r"\b(19|20)\d{2}%k(19|20)\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_incorrect_datewrite(p):
```

Comment:
It could be good to add a negative value for both `-s` and `-d`. Please remember that you can feel free to add new cases if you see the need.